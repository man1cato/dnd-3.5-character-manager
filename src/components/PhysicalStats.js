import React from 'react';

const PhysicalStats = ({hp, ac, speed, initiative, onInputChange}) => (     
    <div>
        <h3 className="text-center mb-5">Physical Stats</h3>
        <div className="input-group--stats">
            <h4>HP</h4>

            <div>
                <h5>Base</h5>
                <div>{hp.base}</div>
            </div>

            <div>+</div>

            <div>
                <h5>Mod</h5>
                <input
                    type="text"
                    name="hp"
                    value={hp.mod}
                    onChange={onInputChange}
                />
            </div>

            <div>-</div>

            <div>
                <h5>Dmg</h5>
                <input
                    type="text"
                    name="hp"
                    id="damage"
                    value={hp.damage}
                    onChange={onInputChange}
                />
            </div>
            <div>=</div>
            <div>
                <h5>Curr</h5>
                <div>{hp.total}</div>
            </div>
        </div>

        <div className="input-group--stats">
            <h4>AC</h4>

            <div>
                <h5>Base</h5>
                <div>{ac.base}</div>
            </div>

            <div>
                <h5>Flat</h5>
                <div>{ac.flat}</div>
            </div>
            
            <div>
                <h5>Touch</h5>
                <div>{ac.touch}</div>
            </div>
        </div>

        <div className="input-group--stats">
            <h4>Init</h4>
            
            <div>
                <h5>Base</h5>
                <div>{initiative.base}</div>
            </div>
            
            <div>+</div>
            
            <div>
                <h5>Mod</h5>
                <input
                    type="text"
                    name="initiative"
                    value={initiative.mod}
                    onChange={onInputChange}
                />
            </div>
            
            <div>=</div>
            
            <div>
                <h5>Total</h5>
                <div>{initiative.total}</div>
            </div>                
        </div>

         <div className="input-group--stats">
            <h4>Speed</h4>
            {speed.ground ? (
                <React.Fragment>   
                    <div>
                        <h5>Ground</h5>
                        <div>{speed.ground}</div>
                    </div>
                    <div>
                        <h5>Flight</h5>
                        <div>{speed.flight}</div>
                    </div>                 
                </React.Fragment>
            ) : (
                <div>{speed} ft/round</div>            
            )}         
         </div>


        
    </div>

)

export default PhysicalStats;