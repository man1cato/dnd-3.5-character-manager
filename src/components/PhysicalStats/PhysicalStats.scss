@import '../../styles/base/settings';


.PhysicalStats__row {
   display: flex;
   > :first-child {
      flex: 0 1 25%;
      align-self: flex-end;
      padding-right: $s-size;
   }
   > :last-child {
      flex: 1 1 75%;
      display: flex;
      align-items: stretch;
      justify-content: space-between;
   }
   input {
      max-width: 40px;
      text-align: center;
   }
   margin-bottom: $m-size;
}

.PhysicalStats__row__item {
   display: flex;
   flex-direction: column;
   align-items: center;
   justify-content: space-between;
   > * {
      margin-top: $xs-size
   }
}

.PhysicalStats__row__item--bottom {
   @extend .PhysicalStats__row__item;
   align-self: flex-end;
}