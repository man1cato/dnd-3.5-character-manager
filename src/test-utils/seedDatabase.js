import _ from 'lodash'
import profile from './fixtures/profile'
import database from '../firebase/firebase'


export const characterOne = {
   ..._.omit(profile, ['id'])
}

export const characterTwo = {
   name: 'Korg',
   gender: 'Male',
   race: 'recoDD3gAaQXU3A56',
   jobClass: 'recbxccm8pp3fTmQw',
   alignment: 'Lawful Evil',
   level: 5,
   iconUrl: 'images/orc.svg'
}

export const characterThree = {
   name: 'Rylia',
   gender: 'Female',
   race: 'recFEpQ9zuD9Yqu7l',
   jobClass: 'recQrfCKtksnnSsaZ',
   alignment: 'Chaotic Neutral',
   level: 3,
   iconUrl: 'images/halfling.svg'
}

export const userOne = {
   uid: 'userOne',
   characterIds: ['characterOne', 'characterTwo']
}

export const userTwo = {
   uid: 'userTwo',
   characterIds: ['characterThree']
}


export default async () => {
   await database.ref('users').remove()
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[0]}`).set(characterOne)
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[1]}`).set(characterTwo)
   await database.ref(`users/${userTwo.uid}/profiles/${userTwo.characterIds[0]}`).set(characterThree)
}