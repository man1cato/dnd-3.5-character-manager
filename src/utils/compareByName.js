export default (a,b) => {
    const aName = a.name.toUpperCase();
    const bName = b.name.toUpperCase();
    if (aName < bName) { return -1 };
    if (aName > bName) { return 1 };
    return 0;
}