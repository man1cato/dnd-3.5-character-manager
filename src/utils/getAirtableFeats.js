import dotenv from 'dotenv'
dotenv.config()

import axios from 'axios'
import _ from 'lodash'

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'

export default async () => {
	try {
		let offset
		let feats = []

		do {
			const res = await axios.get(`${baseUrl}/Feats?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			const featGroup = res.data.records.map((feat) =>  ({
				id: feat.id,
				name: feat.fields.Name,
				type: feat.fields.Type.join(", ") || null,
				description: feat.fields.Description || null
			}));
			
			feats = feats.concat(featGroup)
		} while (!!offset)
		
		return _.orderBy(feats, ['name'], ['asc'])
	} catch (e) {
		console.error(e)
	}
}