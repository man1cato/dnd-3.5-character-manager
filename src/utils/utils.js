import _ from 'lodash'


export const convertTextToArray = (string) => _.compact(string.split(/\n/))

export const findItemById = (storeItemArray, itemId) => storeItemArray.find((item) => item.id === itemId)

export const apiObjectToArray = (obj) => _.orderBy(Object.keys(obj).map((id) => ({
	id,
	...obj[id]
})), ['name'], ['asc'])