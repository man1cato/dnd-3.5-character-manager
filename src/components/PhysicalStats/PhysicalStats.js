import React, { useState, useEffect } from 'react'

import { convertInputValue } from '../../utils/utils'
import './PhysicalStats.scss'


const PhysicalStats = props => {
	const { hp, ac, initBase, speed, handleUpdate } = props
	const [hpMod, setHpMod] = useState(hp.mod || 0)
	const [hpDamage, setHpDamage] = useState(hp.damage || 0)
	const [hpTotal, setHpTotal] = useState(hp.base + hpMod + hpDamage)

	const [initMod, setInitMod] = useState(props.initMod || 0)
	const [initTotal, setInitTotal] = useState(initBase + initMod)

	useEffect(() => {
		setHpTotal(hp.base + hpMod - hpDamage)
		handleUpdate({
			hp: {
				base: hp.base,
				mod: hpMod,
				damage: hpDamage
			}
		})
	}, [hpMod, hpDamage])

	useEffect(() => {
		setInitTotal(initBase + initMod)
		handleUpdate({
			initiativeMod: initMod
		})
	}, [initMod])

	return (
		<>
			<div className="PhysicalStats__row">
				<div>Hit Points</div>
				<div>
					<div className="PhysicalStats__row__item">
						<h5>Base</h5>
						<div>{hp.base}</div>
					</div>
					<div className="PhysicalStats__row__item--bottom">+</div>
					<div className="PhysicalStats__row__item">
						<h5>Mod</h5>
						<input
							type="number"
							data-testid="hpMod"
							value={convertInputValue(hpMod)}
							onChange={(e) => setHpMod(Number(e.target.value))}
						/>
					</div>
					<div className="PhysicalStats__row__item--bottom">-</div>
					<div className="PhysicalStats__row__item">
						<h5>Dmg</h5>
						<input
							type="number"
							data-testid="hpDamage"
							value={convertInputValue(hpDamage)}
							onChange={(e) => setHpDamage(Number(e.target.value))}
						/>
					</div>
					<div className="PhysicalStats__row__item--bottom">=</div>
					<div className="PhysicalStats__row__item">
						<h5>Curr</h5>
						<div data-testid="hpTotal">{hpTotal}</div>
					</div>
				</div>
			</div>

			<div className="PhysicalStats__row">
				<div>Armor Class</div>
				<div>            
					<div className="PhysicalStats__row__item">
						<h5>Base</h5>
						<div>{ac.base}</div>
					</div>
					<div className="PhysicalStats__row__item">
						<h5>Flat</h5>
						<div>{ac.flat}</div>
					</div>
					<div className="PhysicalStats__row__item">
						<h5>Touch</h5>
						<div>{ac.touch}</div>
					</div>
				</div>
			</div>

			<div className="PhysicalStats__row">
				<div>Initiative</div>
				<div>            
					<div className="PhysicalStats__row__item">
						<h5>Base</h5>
						<div>{initBase}</div>
					</div>
					<div className="PhysicalStats__row__item--bottom">+</div>
					<div className="PhysicalStats__row__item">
						<h5>Mod</h5>
						<input
							type="number"
							data-testid="initMod"
							value={convertInputValue(initMod)}
							onChange={(e) => setInitMod(Number(e.target.value))}
						/>
					</div>
					<div className="PhysicalStats__row__item--bottom">=</div>
					<div className="PhysicalStats__row__item">
						<h5>Total</h5>
						<div>{initTotal}</div>
					</div>                
				</div>
			</div>
	
			<div className="PhysicalStats__row">
				<div>Speed</div>
				{speed.ground ? 
					<div>   
						<div className="PhysicalStats__row__item">
							<h5>Ground</h5>
							<div>{speed.ground}</div>
						</div>
						<div className="PhysicalStats__row__item">
							<h5>Flight</h5>
							<div>{speed.flight}</div>
						</div>                 
					</div>
				:
					<div>{speed} ft/round</div>            
				}       
			</div>
		</>
	)
}


export default PhysicalStats