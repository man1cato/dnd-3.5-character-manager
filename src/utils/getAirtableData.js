import axios from 'axios'
import _ from 'lodash'
import { convertTextToArray } from './utils' 
import dotenv from 'dotenv'
dotenv.config({ path: './config/dev.env' })

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'


export const getFeats = async () => {
	try {
		let offset
		let feats = {}

		do {
			const res = await axios.get(`${baseUrl}/Feats?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((feat) => {
				feats[feat.id] = {
					name: feat.fields.Name,
					types: feat.fields.Types || null,
					description: feat.fields.Description || null,
					prerequisites: {
						level: feat.fields['Prereq. Level'] || null,
						class: feat.fields['Prereq. Class'] || null,
						feats: feat.fields['Prereq. Feats'] || null
					} || null
				}
			})			
		} while (!!offset)
		
		return feats
	} catch (e) {
		console.error(e)
	}
}


export const getItems = async () => {
	try {
		let offset
		let items = {}
		do {
			const res = await axios.get(`${baseUrl}/Items?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
			offset = res.data.offset
			res.data.records.forEach((item) => {
				items[item.id] = {
					name: item.fields.Name,
					value: item.fields["Value (gp)"] || '',
					weight: item.fields["Weight (lbs)"] || '',
					category: item.fields.Category || null,
					weaponType: item.fields["Weapon Type"] || null,
					encumbrance: item.fields.Encumbrance || null,
					damageType: item.fields["Damage Type"] ? item.fields["Damage Type"].join(" / ") : [],
					damageS: item.fields["Damage (S)"] || null,
					damageM: item.fields["Damage (M)"] || null,
					critical: item.fields.Critical || null,
					range: item.fields.Range || null,
					armorBonus: item.fields["Armor Bonus"] || null,
					maxDexBonus: item.fields["Max DEX Bonus"] || null
				}
			})
		} while (!!offset)
		
		return items
	} catch (e) {
		console.error(e)
	}
}


const getLevels = async (jobClassId) => {
	try {
		const levelRes = await axios.get(`${baseUrl}/Levels?filterByFormula={Class ID}="${jobClassId}"&api_key=${apiKey}`)
		
		const levelRecords = levelRes.data.records
		let levels = {}
		for (let i = 0; i < levelRecords.length; i++) {
			const levelRecord = levelRecords[i].fields 
			levels[levelRecord.Level] = {
				saves: {
					fortitude: levelRecord['Fortitude'],
					reflex: levelRecord['Reflex'],
					will: levelRecord['Will']
				},
				baseAttackBonuses: _.compact([ levelRecord['BAB 1'], levelRecord['BAB 2'], levelRecord['BAB 3'], levelRecord['BAB 4'] ]),
				xp: levelRecord['XP Required'],
				specialAbilities: !!levelRecord['Special Abilities'] ? levelRecord['Special Abilities'] : null,
				spellsPerDay:  _.compact([ 
					levelRecord['SPD 0'], levelRecord['SPD 1'], levelRecord['SPD 2'], levelRecord['SPD 3'], levelRecord['SPD 4'], 
					levelRecord['SPD 5'], levelRecord['SPD 6'], levelRecord['SPD 7'], levelRecord['SPD 8'], levelRecord['SPD 9'] 
				])
			}
		}
		return levels
	} catch (e) {
		console.error(e)
	}
}

export const getJobClasses = async () => {
	try {
		let offset
		let jobClasses = {}
	
		do {
			const classRes = await axios.get(`${baseUrl}/Classes?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = classRes.data.offset

			const jobClassGroup = classRes.data.records
			
			for (let i = 0; i < jobClassGroup.length; i++) {
				const jobClass = jobClassGroup[i] 
				jobClasses[jobClass.id] = {
					name: jobClass.fields.Name,
					hitDie: jobClass.fields['Hit Die'],
					proficiencies: jobClass.fields['Proficiencies'] || [],
					skills: jobClass.fields.Skills || [],
					levels: await getLevels(jobClass.id)
				}
			}
		} while (!!offset)
		
		return jobClasses
	} catch (e) {
		console.error(e)
	}
}


export const getRaces = async () => {
	try {
		let offset
		let races = {}
		do {
			const res = await axios.get(`${baseUrl}/Races?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((race) =>  {
				races[race.id] = {
					name: race.fields.Name,
					size: race.fields.Size,
					speed: race.fields['Speed (ft/rnd)'],
					racialBonuses: race.fields['Racial Bonuses'] ? convertTextToArray(race.fields['Racial Bonuses']) : null,
					defaultLanguages: race.fields['Languages (Default)'],
					bonusLanguages: race.fields['Languages (Bonus)'],
					abilityMods: {
						str: race.fields['STR Mod'],
						dex: race.fields['DEX Mod'],
						con: race.fields['CON Mod'],
						int: race.fields['INT Mod'],
						wis: race.fields['WIS Mod'],
						cha: race.fields['CHA Mod']
					},
					specialAbilities: race.fields['Special Abilities'] || [],
					favoredClass:  race.fields['Favored Class'] ? race.fields['Favored Class'][0] : null,
					iconUrl: race.fields['Icon URL']
				}
			})
			
		} while (!!offset)
		
		return races
	} catch (e) {
		console.error(e)
	}
}


export const getSkills = async () => {
   let offset
   let skills = {}
   do {
		const res = await axios.get(`${baseUrl}/Skills?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
		offset = res.data.offset
		res.data.records.forEach((skill) => {
			skills[skill.id] = {
				name: skill.fields.Name,
				keyAbility: skill.fields["Key Ability"] || null,
				description: skill.fields.Description || "",
				notes: skill.fields.Notes || ""
			}
		})
   } while (!!offset)
   
   return skills
}


export const getSpecialAbilities = async () => {
   let offset
   let specialAbilities = {}
   do {
		const res = await axios.get(`${baseUrl}/Special Abilities?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
		offset = res.data.offset
		res.data.records.forEach((specialAbility) => {
			specialAbilities[specialAbility.id] = {
				name: specialAbility.fields.Name,
				type: specialAbility.fields.Type ? specialAbility.fields.Type.join(", ") : null,
				description: specialAbility.fields.Description				
			}
		})
   } while (!!offset)
   
   return specialAbilities
}


export const getSpells =  async () => {
   let offset
   let spells = {}
   do {
		const res = await axios.get(`${baseUrl}/Spells?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '')) 
		offset = res.data.offset
		res.data.records.forEach((spell) => {
			spells[spell.id] = {
				name: spell.fields.Name,
				description: spell.fields.Description || "",
				school: spell.fields.School || null,
				components: spell.fields.Components.join(", ") || null,
				castingTime: spell.fields["Casting Time"] || null,
				range: spell.fields.Range || null,
				effectArea: spell.fields["Target, Effect, Area"] || null,
				duration: spell.fields.Duration || null,
				savingThrow: spell.fields["Saving Throw"] || null,
				spellResistance: spell.fields["Spell Resistance"] || null
			}
		})
   } while (!!offset)

   return spells
}