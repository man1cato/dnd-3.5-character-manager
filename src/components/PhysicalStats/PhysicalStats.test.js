import React from 'react'
import { mount } from 'enzyme'

import PhysicalStats from './PhysicalStats'
import { characterOne } from '../../tests/utils/seedDatabase'
import { apiData } from '../../tests/utils/utils'

const hp = characterOne.hp
const handleUpdate = jest.fn()
let props, wrapper

beforeAll(async () => {
    const api = await apiData()
    props = {
        hp,
        ac: characterOne.ac,
        initBase: 4,
        initMod: characterOne.initMod,
        speed: api.races[characterOne.race].speed,
        handleUpdate
    }
})

beforeEach(() => {
    wrapper = mount(<PhysicalStats {...props} />)
})

test('should render attacks with default profile data', () => {
    expect(wrapper).toMatchSnapshot()
})

test('should trigger handleUpdate when input changes', () => {
    const value = 2
	wrapper.find('#hpMod').simulate('change', { target: { value } })
    expect(handleUpdate).toHaveBeenCalledWith({
        hp: { ...hp, mod: value }
    })
})