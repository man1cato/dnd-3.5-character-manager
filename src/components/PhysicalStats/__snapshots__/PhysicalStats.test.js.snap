// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render attacks with default profile data 1`] = `
<PhysicalStats
  ac={
    Object {
      "base": 14,
      "flat": 10,
      "touch": 14,
    }
  }
  handleUpdate={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "hp": Object {
              "base": 32,
              "damage": 10,
              "mod": 6,
            },
          },
        ],
        Array [
          Object {
            "initiativeMod": 2,
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  hp={
    Object {
      "base": 32,
      "damage": 10,
      "mod": 6,
    }
  }
  initBase={4}
  initMod={2}
  speed={30}
>
  <div
    className="PhysicalStats__row"
  >
    <div>
      Hit Points
    </div>
    <div>
      <div
        className="PhysicalStats__row__item"
      >
        <h5>
          Base
        </h5>
        <div>
          32
        </div>
      </div>
      <div
        className="PhysicalStats__row__item--bottom"
      >
        +
      </div>
      <div
        className="PhysicalStats__row__item"
      >
        <h5>
          Mod
        </h5>
        <input
          id="hpMod"
          onChange={[Function]}
          type="number"
          value={6}
        />
      </div>
      <div
        className="PhysicalStats__row__item--bottom"
      >
        -
      </div>
      <div
        className="PhysicalStats__row__item"
      >
        <h5>
          Dmg
        </h5>
        <input
          id="hpDamage"
          onChange={[Function]}
          type="number"
          value={10}
        />
      </div>
      <div
        className="PhysicalStats__row__item--bottom"
      >
        =
      </div>
      <div
        className="PhysicalStats__row__item"
      >
        <h5>
          Curr
        </h5>
        <div>
          28
        </div>
      </div>
    </div>
  </div>
  <div
    className="PhysicalStats__row"
  >
    <div>
      Armor Class
    </div>
    <div>
      <div
        className="PhysicalStats__row__item"
      >
        <h5>
          Base
        </h5>
        <div>
          14
        </div>
      </div>
      <div
        className="PhysicalStats__row__item"
      >
        <h5>
          Flat
        </h5>
        <div>
          10
        </div>
      </div>
      <div
        className="PhysicalStats__row__item"
      >
        <h5>
          Touch
        </h5>
        <div>
          14
        </div>
      </div>
    </div>
  </div>
  <div
    className="PhysicalStats__row"
  >
    <div>
      Initiative
    </div>
    <div>
      <div
        className="PhysicalStats__row__item"
      >
        <h5>
          Base
        </h5>
        <div>
          4
        </div>
      </div>
      <div
        className="PhysicalStats__row__item--bottom"
      >
        +
      </div>
      <div
        className="PhysicalStats__row__item"
      >
        <h5>
          Mod
        </h5>
        <input
          id="initMod"
          onChange={[Function]}
          type="number"
          value={2}
        />
      </div>
      <div
        className="PhysicalStats__row__item--bottom"
      >
        =
      </div>
      <div
        className="PhysicalStats__row__item"
      >
        <h5>
          Total
        </h5>
        <div>
          6
        </div>
      </div>
    </div>
  </div>
  <div
    className="PhysicalStats__row"
  >
    <div>
      Speed
    </div>
    <div>
      30
       ft/round
    </div>
  </div>
</PhysicalStats>
`;
